packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,139 @@
+
+  /**
+   * Unbinds the given observer from the component associated with this Producer.
+   * @param consumer - The observer to unregister from the Producer.
+   */
+  removeConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Return IDataResult for the given dataBindingReference, containing the data value
+   * and some additional metadata.
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
+
+  /**
+   * Invoked when this data provider is deleted.
+   */
+  onDeleted(): void;
+}
+
+/**
+ * The interface for an object that can bind to an IProducer.
+ * One example of this provider-observer relationship is a Formula binding
+ * to Tablero. In this scenario, Formula is an IConsumer listening to changes
+ * in the IProducer, Tablero.
+ *
+ * Any object that implements IConsumer is expected to provide a
+ * callback whenever the component it is bound to changes in value and a reference
+ * to the data that the observer is bound to.
+ */
+export interface IConsumer<T> {
+  /**
+   * Invoked whenever the data this object is bound to is changed.
+   */
+  notify: <K extends keyof T>(property: K, value: T[K]) => void;
+
+  /**
+   * Invoked whenever the provider this is bound to is deleted.
+   */
+  onProducerDeleted: () => void;
 }
 
-export type PickPending<T, K extends keyof T> = { [Key in K]: T[K] | Pending<T[K]> };
-
-export interface Producer<T> {
-    id: string;
-    unsubscribe: (origin: Consumer<T>) => void;
-    
-    now: <K extends keyof T>(property: K, ...args: any[]) => T[K] | undefined;
-    nowMany: <K extends keyof T>(properties: Record<K, undefined | any[]>) => Partial<Pick<T, K>>;
-    
-    request: <K extends keyof T>(origin: Consumer<T>, property: K, ...args: any[]) => T[K] | Pending<T[K]>;
-    requestMany: <K extends keyof T>(origin: Consumer<T>, properties: Record<K, undefined | any[]>) => PickPending<T,K>;
+/**
+ * IDataResult represents the result object passed through data binding between two components.
+ * It provides query functions for retrieving values, hints, friendly names, and provides
+ * basic traverse APIs to move to previous or next level results.
+ */
+export interface IDataResult {
+  /**
+   * Get the list of properties of this dataResult.
+   * @returns a list of dataResults representing the next level objects.
+   */
+  getProperties: (query: string) => Promise<IDataResult[]>;
+
+  /**
+   * Return the raw value of this object.
+   */
+  getValue: () => Pending<IData | undefined>;
+
+  /**
+   * Return the hint text of this object.
+   */
+  getHint: () => string | undefined;
+
+  /**
+   * Return the friendlyName of this object.
+   */
+  getFriendlyName: () => string;
+
+  /**
+   * Return The dataBindingReference for this instance of Producer.
+   */
+  getBindingReference: () => any;
```

Comment:
I think getBindingReference is just the key used to read the value?

Reply:
Sort of. Currently we have the Consumer hold on to the IDataResult that it's bound to. Since we can't store it over the wire, the binding reference is a serializable object that the Producer knows know how to interpret to find the right IDataResult.

I think it makes more sense though for the Producer to keep track of what values the Consumers are bound to. We had been planning to redesign our use of data binding to get rid of this function so I'm all for removing it